docs/adrs/0079-api-versioning.md

```diff
@@ -0,0 +1,124 @@
+  * Cons: More generated files to manage
+
+**Chosen Alternative:** Option 2. While it does increase the number of files, it will be clearer to differentiate and easier to maintain a separate swagger file and its accompanying generated files. [You can see an example in this commit](https://github.com/transcom/mymove/pull/10816/commits/0f387d655e098e589b10ab27b388bc771e873f09).
+
+#### Handler Organization
+
+Since we will be generating separate `*messages` and `*operations` packages for the version 2 api, it would make sense then to create a separate directory for the version 2 handlers. This will make it easier to differentiate the handlers from different versions, and ensure we are using the correct generated code for that handler.
+
+You can see an example of utilizing the generated code in the handlers and creating the necessary routing [in this commit](https://github.com/transcom/mymove/pull/10816/commits/0c4be3bab159a49f83e407830a9fecb9d27278da).
+
+#### Services/Fetcher Organization
+
+This is where there are a few options to organize our services. And I am currently unsure about the best way to go.
+
+Here are some questions to consider, when adding a new endpoint to our version 2 API:
+
+1. What do we do if we create a new endpoint that requires changes to the service that is called? How do we organize the functions/files? Do we create a new file in the services folder? Should we store these services files in a separate v2 subdir? What is the naming structure going to look like?
+2. What if the new endpoint does not require changes to the service? Should we make a new service anyway to prevent issues further down the line?
+
+
+
+* Option 1: Rename the old version to something that indicates it is old and add the new version to the same fetcher.
+  * Pros: As we remove the deprecated endpoints, we will have less code duplication.
+  * Cons: It will be easier to use the wrong service with the wrong endpoint.
+  * [Example can be seen here](https://github.com/transcom/mymove/pull/10816/commits/0913f2f4625450a2ac43b358e1890d17d77b747d)
+* Option 2: Create a new fetcher and service for the new version. This can be separated out into a new file or put in the same services file. In the example linked below it is separated out into its own file.
+  * Pros: Clear delineation between the two versions. You would not be able to mix up the new service with the old service since it does not use the same fetcher.
+  * Cons: Repetitive code. Naming questions arise, in order to delineate the new from the old. Difficulty when dealing with more complicated services, especially when we use a service in another service.
+  * [Example can be seen here](https://github.com/transcom/mymove/compare/0c4be3b...3769845)
+
+There are definitely other alternatives out there. Please feel free to suggest another.
```

Comment:
**thought**: So, this idea just off the top of my head and tbh, I have no idea if it would actually work. What if we approached this a bit like how version managers approach things in terms of:
* something knows what version of the api/service we want (maybe a var we keep in context)
* every service has a top level entry point, e.g. `MoveTaskOrderFetcher.ListPrimeMoveTaskOrders`, which serves more as a router based on the version, and calls the appropriate version of the service
* versions of a given service are subdirectories of the service package

So for example, with the `MoveTaskOrderFetcher`, the directory structure would be something like:

```
pkg/
    move/
        move_fetcher/
            main.go    <- shell/router service which contains `MoveTaskOrderFetcher` and calls the appropriate service version
            v1/
                main.go <- contains `MoveTaskOrderFetcherV1`
            v2/
                main.go <- contains `MoveTaskOrderFetcherV2`
```

It would maybe be a bit of re-work in terms of how the services are initialized, but tbh I'm not sure. If it doesn't, this could be a pattern that can be adopted incrementally as a given service has the need to change versions. 

In this pattern, I wonder if we could somehow set the internal/ghc API request context to always have the latest version or something 🤔

Reply:
@felipe-lee This suggestion is interesting. It seems like option 2, but it keeps the services even more separate and distinct. I will see if I can make a commit that illustrates this idea on my test branch.

Reply:
**thought:** After thinking about it more, the subdir sounds like the best option to me, so similar to what Felipe recommended above. 
Having the different version of the endpoint in the same file can cause the file to get rather large and hard to navigate within the file, having a different version of the file in the same directory (without sub directorying) makes the directory cluttered and also hard to navigate. The only thing that would be problematic here would be having to be careful about imports and making sure folks who write code are importing from the correct version.